staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,6 +69,251 @@ func TestSimpleQueue(t *testing.T) {
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 1*time.Second, AddAfterOptions{Existing: TakeLonger, Synchronous: true})
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// Add a longer delay with the expectation of pushing the ready later...
+	q.AddAfterWithOptions(second, 20*time.Second, AddAfterOptions{Existing: TakeLonger, Synchronous: true})
+
+	// step past the original delayed item and check it queues.
+	fakeClock.Step(4 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// step past the new longer delay and prove our longer item was queued.
+	fakeClock.Step(30 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+
+	item, _ = q.Get()
+	q.Done(item)
+
+	// Show that TakeIncoming will override whatever the value of the waiting item
+	// Shorter case...
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 1*time.Second, AddAfterOptions{Existing: TakeIncoming, Synchronous: true})
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+	item, _ = q.Get()
+	q.Done(item)
+
+	// Longer case...
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 20*time.Second, AddAfterOptions{Existing: TakeIncoming, Synchronous: true})
+	// step past the first waiting time...
+	fakeClock.Step(10 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+}
+
+// TestForgetDelayed - should remove an item from the delayed queue
+func TestForgetDelayed(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewDelayingQueueWithCustomClock(fakeClock, "")
+
+	first := "foo"
+
+	// add the item into the waiting queue.
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// forget the item
+	q.ForgetDelayed(first, ForgetDelayedOptions{Synchronous: true})
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// step past the original delayed item and ensure it no longer exists
+	fakeClock.Step(7 * time.Second)
```

Comment:
we should be testing by manipulating the clock and what gets received to the queue.

You can check or copy tests from my old [PR](https://github.com/kubernetes/kubernetes/pull/113607/files#diff-500d9650a8786e56a89aef21e37488ae830cd3b4334d8075e2ab297730b0f748) if you want to enhance yours. I think we should at least have more items, to assert we removed the correct one. We could also test that calling ForgetDelayed does not remove the item when it is already in the queue.

Reply:
Ok have now done this.  I've used a new AssertState test helper more thoroughly assert the state of the two queues we expect and cut down on boiler plate code in the tests.